src/Conversion/ONNXToKrnl/Math/LRN.cpp

```diff
@@ -55,87 +55,86 @@ struct ONNXLRNOpLowering : public ConversionPattern {
-    SmallVector<Value, 4> loadIndices;
-    for (int i = 0; i < outputRank; i++) {
-      if (i == loopIndexForC) {
-        Value loopVal = sumLoops.getInductionVar(0);
-        loadIndices.emplace_back(loopVal);
-      } else {
-        Value loopVal = outputLoops.getInductionVar(i);
-        loadIndices.emplace_back(loopVal);
-      }
-    }
-
-    Value loadVal = create.krnl.load(input, loadIndices);
-    Value squareVal = create.math.mul(loadVal, loadVal);
-
-    Value sumValue = create.krnl.load(sumAlloc, ArrayRef<Value>{});
-    sumValue = create.math.add(sumValue, squareVal);
-    create.krnl.store(sumValue, sumAlloc, ArrayRef<Value>{});
-
-    // Compute and store the output
-    // y = x / ((bias + (alpha / nsize) * square_sum) ** beta)
-    rewriter.restoreInsertionPoint(outputLoopBody);
-    SmallVector<Value, 4> storeIndices;
-    for (int i = 0; i < outputRank; ++i) {
-      storeIndices.emplace_back(outputLoops.getInductionVar(i));
-    }
-    Value xValue = create.krnl.load(input, storeIndices);
-    sumValue = create.krnl.load(sumAlloc);
-    Value tempValue =
-        create.math.pow(create.math.add(biasValue,
-                            create.math.mul(alphaDivSizeValue, sumValue)),
-            betaValue);
-    Value resultValue = create.math.div(xValue, tempValue);
-
-    create.krnl.store(resultValue, alloc, storeIndices);
+    KrnlBuilder createKrnl(rewriter, loc);
+    ValueRange outputLoopDef = createKrnl.defineLoops(outputRank);
+    SmallVector<IndexExpr, 4> lbs(outputRank, LiteralIndexExpr(0));
+    createKrnl.iterateIE(outputLoopDef, outputLoopDef, lbs,
+        shapeHelper.dimsForOutput(0),
+        [&](KrnlBuilder &createKrnl, ValueRange outputLoopInd) {
+          // Insert computation of square_sum.
+          // square_sum[n, c, d1, ..., dk] = sum(X[n, i, d1, ..., dk] ^ 2),
+          // where max(0, c - floor((size - 1) / 2)) <= i
+          // and i<= min(C - 1, c + ceil((size - 1) / 2)).
+
+          // Get a child IndexExpr context.
+          IndexExprScope childScope(&rewriter, shapeHelper.scope);
+
+          // Compute the lower bound and upper bound for square_sum.
+          const int loopIndexForC = 1;
+          DimIndexExpr cIE(outputLoopInd[loopIndexForC]);
+          MemRefBoundsIndexCapture inputBounds(input);
+          DimIndexExpr CIE(inputBounds.getDim(loopIndexForC));
+          SymbolIndexExpr sizeIE = LiteralIndexExpr(sizeLit);
+
+          SmallVector<IndexExpr, 2> lbMaxList;
+          lbMaxList.emplace_back(LiteralIndexExpr(0));
+          lbMaxList.emplace_back(
+              cIE - (sizeIE - 1).floorDiv(LiteralIndexExpr(2)));
+
+          SmallVector<IndexExpr, 2> ubMinList;
+          ubMinList.emplace_back(CIE);
+          ubMinList.emplace_back(
+              cIE + 1 + (sizeIE - 1).ceilDiv(LiteralIndexExpr(2)));
+
+          // Initialize sum, single scalar, no need for default alignment.
+          MemRefType scalarMemRefType = MemRefType::get({}, elementType, {}, 0);
```

Comment:
What is the zero in `MemRefType::get({}, elementType, {}, 0);` ? 
Suggest to use: 
```
MemRefType scalarMemRefType = MemRefType::get({}, elementType);
```
which should be equivalent.

Reply:
`MemRefType::get({}, elementType)` invokes:
```
static MemRefType get(ArrayRef<int64_t> shape, Type elementType, MemRefLayoutAttrInterface layout = {}, Attribute memorySpace = {});

MemRefType MemRefType::get(ArrayRef<int64_t> shape, Type elementType,                                                                      
                           MemRefLayoutAttrInterface layout,                                                                               
                           Attribute memorySpace) {                                                                                        
  // Use default layout for empty attribute.
  if (!layout)
    layout = AffineMapAttr::get(AffineMap::getMultiDimIdentityMap(
        shape.size(), elementType.getContext()));                                                                                          
  
  // Drop default memory space value and replace it with empty attribute.                                                                  
  memorySpace = skipDefaultMemorySpace(memorySpace);                                                                                       
                                                                                                                                           
  return Base::get(elementType.getContext(), shape, elementType, layout,                                                                   
                   memorySpace);
} 
```
`MemRefType::get({}, elementType, {}, 0)` invokes:
```
static MemRefType get(ArrayRef<int64_t> shape, Type elementType, AffineMap map, unsigned memorySpaceInd);

MemRefType MemRefType::get(ArrayRef<int64_t> shape, Type elementType,
                           AffineMap map, unsigned memorySpaceInd) {

  // Use default layout for empty map.
  if (!map)
    map = AffineMap::getMultiDimIdentityMap(shape.size(),
                                            elementType.getContext());

  // Wrap AffineMap into Attribute.
  Attribute layout = AffineMapAttr::get(map);

  // Convert deprecated integer-like memory space to Attribute.
  Attribute memorySpace =
      wrapIntegerMemorySpace(memorySpaceInd, elementType.getContext());

  return Base::get(elementType.getContext(), shape, elementType, layout,
                   memorySpace);
}
```

Reply:
ok the implementation differs so you cannot simplify the code. Leave it as is.